docs: SNP attestation documentation (#1333)
edgelesssys/contrast: docs/docs/architecture/snp.md

# SNP Attestation

> The key components for attesting AMD SEV-SNP are the Security Processor (SP),

Reviewer: ```suggestion
The key component for attesting AMD SEV-SNP is the Security Processor (SP),
```

| SIGNATURE    | Signature of bytes `0h` to `29Fh` inclusive of this report.            | Used to verify the integrity and authenticity of the report.

### Platform Info Structure
The platform info structure is embedded in the attestation report in the `PLATFORM_INFO` field.
The complete structure can be found in the SEV Secure Nested Paging Firmware ABI Specification in table 24.

| Field    | Description  |
| -------- | -------    |
| ALIAS_CHECK_COMPLETE | Indicates that alias detection has completed since the last system reset and there are no aliasing addresses. Resets to 0. Contains mitigation for CVE-2024-21944.        |  |
| CIPHERTEXT_HIDING_DRAM_EN | Indicates ciphertext hiding is enabled for the DRAM.        |  |
| RAPL_DIS | Indicates that the RAPL feature is disabled.        |  |
| ECC_EN | Indicates that the platform is using error correcting codes for memory. Present when EccMemReporting feature bit is set.        |  |
| TSME_EN | Indicates that TSME is enabled in the system.        |  |
| SMT_EN | Indicates that SMT is enabled in the system.        |  |

## Anonymous ID Block Signing

> To both have the ability to sign ID Blocks and not be part of the TCB, we must ensure

Reviewer 2: needs more motivation/background

Author: See https://github.com/edgelesssys/contrast/pull/1333/commits/2e56a478054edceb907cddd62e44cddfaf08423a

that there exists no private key for the `ID_KEY` in the ID Auth structure.
For this, we implement ECDSA public key recovery. The algorithm is defined in [SEC 1: Elliptic Curve Cryptography](https://www.secg.org/sec1-v2.pdf).

> It allows to calculate a ECDSA public key based given a signature and the message

Reviewer: ```suggestion
It allows to calculate an ECDSA public key given a signature and the message
```

> that was signed. We keep the signature constant as (r,s) = (2,1) for all versions,

Reviewer: ```suggestion
that was signed. We keep the signature constant as `(r,s) = (2,1)` for all versions and
```

use the given ID Block containing the policy and launch digest as an input.

> The recovery algorithm returns 2 valid public keys from which we choose the smaller one.

Reviewer: Might make sense to define "smaller" for curve points.

> The recovery algorithm returns 2 valid public keys from which we choose the smaller one.

Reviewer: ```suggestion
The recovery algorithm returns two valid public keys from which we choose the smaller one.
```
